### PR TITLE
[Transaction] Transaction admin api get slow transaction metadata.

### PR DESCRIPTION
**Pull request by congbobo184**
## Motivation
Transaction add admin api `getSlowTransactions`
Transaction add admin api `getSlowTransactionsByCoordinatorId`

## implement
@Data
public class TransactionMetadata {

    /** The txnId of this transaction. */
    public String txnId;

    /** The status of this transaction. */
    public String status;

    /** The open time of this transaction. */
    public long openTimestamp;

    /** The timeout of this transaction. */
    public long timeoutAt;

    /** The producedPartitions of this transaction. */
    public Map<String, TransactionInBufferStats> producedPartitions;

    /** The ackedPartitions of this transaction. */
    public Map<String, Map<String, TransactionInPendingAckStats>> ackedPartitions;
}
### Verifying this change
Add the tests for it

Does this pull request potentially affect one of the following parts:
If yes was chosen, please highlight the changes

Dependencies (does it add or upgrade a dependency): (no)
The public API: (no)
The schema: (no)
The default values of configurations: (no)
The wire protocol: (no)
The rest endpoints: (no)
The admin cli options: (yes)
Anything that affects deployment: (no)

### CODE REVIEW

**Comment by congbobo184** [comment]
@Anonymitaet please review doc thanks.

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java [hunk 1]
```diff
@@ -95,11 +97,34 @@ void run() throws Exception {
         }
     }
 
+    @Parameters(commandDescription = "Get slow transaction metadata")
+    private class GetSlowTransactionMetadata extends CliCommand {
```
Comment (codelipenghui) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java:101]:
```suggestion
    private class GetSlowTransactions extends CliCommand {
```

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java [hunk 2]
```diff
@@ -95,11 +97,34 @@ void run() throws Exception {
...
+        }
+    }
+
     public CmdTransactions(Supplier<PulsarAdmin> admin) {
         super("transactions", admin);
         jcommander.addCommand("coordinator-status", new GetCoordinatorStatus());
         jcommander.addCommand("transaction-in-buffer-stats", new GetTransactionInBufferStats());
         jcommander.addCommand("transaction-in-pending-ack-stats", new GetTransactionInPendingAckStats());
         jcommander.addCommand("transaction-metadata", new GetTransactionMetadata());
+        jcommander.addCommand("slow-transaction-metadata", new GetSlowTransactionMetadata());
```
Comment (codelipenghui) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTransactions.java:128]:
```suggestion
        jcommander.addCommand("slow-transactions", new GetSlowTransactionMetadata());
```

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Transactions.java [hunk 3]
```diff
@@ -71,4 +72,26 @@
...
 
+    /**
+     * Get slow transaction metadata by coordinatorId.
+     *
+     * @param coordinatorId the coordinator id of getting slow transaction status.
+     * @param timeout the timeout
+     * @param timeUnit the timeout timeUnit
+     * @return the future metadata of slow transactions.
+     */
+    CompletableFuture<Map<String, TransactionMetadata>> getSlowTransactionMetadataByCoordinatorId(Integer coordinatorId,
```
Comment (codelipenghui) [inline pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Transactions.java:83]:
```suggestion
    CompletableFuture<Map<String, TransactionMetadata>> getSlowTransactions(Integer coordinatorId,
```

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Transactions.java [hunk 4]
```diff
@@ -71,4 +72,26 @@
...
+
+    /**
+     * Get slow transaction metadata.
+     *
+     * @param timeout the timeout
+     * @param timeUnit the timeout timeUnit
+     *
+     * @return the future metadata of slow transactions.
+     */
+    CompletableFuture<Map<String, TransactionMetadata>> getSlowTransactionMetadata(long timeout, TimeUnit timeUnit);
```
Comment (codelipenghui) [inline pulsar-client-admin-api/src/main/java/org/apache/pulsar/client/admin/Transactions.java:95]:
```suggestion
    CompletableFuture<Map<String, TransactionMetadata>> getSlowTransactions(long timeout, TimeUnit timeUnit);
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 5]
```diff
@@ -286,4 +288,90 @@ protected void internalGetTransactionMetadata(AsyncResponse asyncResponse,
             }
         }
     }
+
+    protected void internalGetSlowTransactionsMetadata(AsyncResponse asyncResponse,
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:292]:
```suggestion
    protected void internalGetSlowTransactions(AsyncResponse asyncResponse,
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 6]
```diff
@@ -286,4 +288,90 @@ protected void internalGetTransactionMetadata(AsyncResponse asyncResponse,
...
+                                                       boolean authoritative, long timeout, Integer coordinatorId) {
+        try {
+            if (pulsar().getConfig().isTransactionCoordinatorEnabled()) {
+                if (coordinatorId != null) {
+                    validateTopicOwnership(TopicName.TRANSACTION_COORDINATOR_ASSIGN.getPartition(coordinatorId),
+                            authoritative);
+                    List<Long> transactions = pulsar().getTransactionMetadataStoreService().getStores()
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:299]: The metadata store might be null.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java [hunk 7]
```diff
@@ -286,4 +288,90 @@ protected void internalGetTransactionMetadata(AsyncResponse asyncResponse,
...
+                            .get(TransactionCoordinatorID.get(coordinatorId)).getSlowTransactions(timeout);
+                    List<CompletableFuture<TransactionMetadata>> completableFutures = new ArrayList<>();
+                    for (Long transaction : transactions) {
+                        completableFutures.add(pulsar().getAdminClient().transactions()
+                                .getTransactionMetadata(new TxnID(coordinatorId, transaction)));
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TransactionsBase.java:304]: Why need to use admin to get the transaction metadata?

File: pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java [hunk 8]
```diff
@@ -419,6 +419,17 @@ public TransactionMetadataStoreStats getStats() {
         return transactionMetadataStoreStats;
     }
 
+    @Override
+    public List<Long> getSlowTransactions(long timeout) {
```
Comment (codelipenghui) [inline pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/MLTransactionMetadataStore.java:423]: Why not return the List<TxnMeta> directly?

**Review by codelipenghui — CHANGES_REQUESTED** [review]

**Review by codelipenghui — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
LGTM

**Pull request merged**